src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+
+It's a restrictive system, but data races are impossible. Huzzah!
+
+
+<slice/>
+
+
+## Why no Mutex?
+
+Having a mutex is not enough to prevent data races.
+
+
+For example, in Java, we would use a mutex like this:
+
+```java
+synchronized (myObj) {
+  Counter counter = myObj.myCounter;
+  myCounter.number = myCounter.number + 1;
+}
+```
+
+This code looks good; the `synchronized` block will lock the mutex in `myObj` [# In Java, every object is (or contains) a mutex.] and then unlock it when the block ends.
+
+
+However, the programmer can still mess this up: they can access `counter` after the block!
+
+```java
+Counter counter;
+synchronized (myObj) {
+  counter = myObj.myCounter;
+  myCounter.number = myCounter.number + 1;
+}
+myCounter.number = myCounter.number + 7;
+```
+
+
+The problem is that references can *outlive* the mutex lock.
+
+
+Luckily, there's a way to prevent that. Enter Rust!
+
+
+<slice/>
+
+
+## Mutexes with Rust
+
+Just like how Pony tracks permissions, Rust tracks lifetimes, and can ensure that no reference escapes a mutex lock.
+
+Here is the equivalent code in Rust: [# [Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8ce5115cc1d4117aeab61da2eef69f03)]
+
+```rust
+let m = Mutex::new(100i64);
+
+let escaped_reference =
+  {
+    let counter = &mut *m.lock().unwrap();
+    *counter = *counter + 1i64;
+    counter
+    // Error, because the counter pointer is escaping!
+  };
+
+println!("{:?}", *escaped_reference);
+```
+
+However, Rust is totally fine letting something else escape, since it didn't come from inside the mutex: [# [Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9151d40e3c0b4b12c72d67754c2a8845)]
+
+
+```rust
+let m = Mutex::new(100i64);
+
+let escaped_value =
+  {
+    let counter = &mut *m.lock().unwrap();
+    *counter = *counter + 1i64;
+    let new_value = *counter + 100
+    new_value
+  };
+
+println!("{:?}", escaped_value);
+```
+
+
+Rust uses the type system to track where a value came from. [# This part of the type system is called the "borrow checker". It's actually much more complicated and powerful than alluded to here.]
+
+
+<slice/>
+
+
+# How do we combine them?
+
+We can make structured concurrency fearless by:
+
+ 1. Start with structured concurrency.
+ 1. We *can't* modify values created *outside* a parallel block.
+ 1. ...except for mutexes, we can always lock them.
+
+
+Our C program _almost_ follows these rules, except it violates rule 2; we modified the `results` array inside the parallel block:
+
+
+```c
+int exponent = 3;
+
+int results[5];
+
+// Launch some threads and run in parallel!
+#pragma omp parallel for
+for (int i = 0; i < 5; i++) {
+  results[i] = pow(i, exponent);
+}
+```
+
+
+
+But Vale has a `foreach` loop that can accumulate each iteration's result, and produce an array: [# For those unfamiliar, Vale is a language that ... (put stuff here)]
+
+
+```
+exponent = 3;
```

Comment:
Put this in a main function, its better for the first code example to be runnable imo

Reply:
Done!